src/types/global.d.ts

```diff
@@ -22,3 +22,9 @@ declare module '*.lottie' {
 }
 
 declare module 'react-native-device-info/jest/react-native-device-info-mock';
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+interface Window {
+    enableMemoryOnlyKeys: () => void;
+    disableMemoryOnlyKeys: () => void;
```

Comment:
I think this requires a comment. We're just extending `Window` without a word.

Reply:
@cubuspl42 Updated!

Reply:
I'm sorry, I wasn't clear enough. This is a top-level app-wise global type definition file. There's no context here.

These are for-some-reason-global functions for enabling/disabling **Onyx** memory only-keys.

https://github.com/Expensify/App/blob/5d45f0f6c86e97d4ae58ac167b6c162a610b817c/src/libs/actions/MemoryOnlyKeys/exposeGlobalMemoryOnlyKeysMethods/index.js#L4-L9

Do I remember correctly that a `.d.ts` file can reside anywhere?

If so, could this expansion be moved to `src/libs/actions/MemoryOnlyKeys/exposeGlobalMemoryOnlyKeysMethods/`? It gives much more context, as the typing would be located next to the functionality it is describing. I think this would be an even better source of context than the comments.

Reply:
> Do I remember correctly that a `.d.ts` file can reside anywhere?

It will go against [this](https://github.com/Expensify/App/blob/main/contributingGuides/TS_STYLE.md#d-ts-extension) rule about using `d.ts` extension.

@fabioh8010 @blazejkustra What do you think?

Reply:
Still, this very rule states that...

> files in which **third party packages can be modified using module augmentation**.

We're adding typing for our own `window` monkey-patching, this is not third-party stuff.

I have no idea why these functions were made global in the first place. Is it so they can be easily invoked from console?... This is weird, there are no comments about that in the .js file.

Reply:
We will not make this _not_ a `.d.ts` file by putting these lines in `types.d.ts`. The rule you quote seems to mean "**Always** prefer .ts extension over .d.ts". We do.

From my perspective, putting first-party typing fully separated from the logic it describes is counter-productive. This is a corner case, though, and the guidelines seem to reason about typical cases.

Reply:
I agree with @VickyStash, it goes against the guidelines. Placing typings fully separated from the logic they describe being counter-productive, it's a valid argument. But we have to balance that with better overall project structure and maintaining consistency across the project. That being said I would leave it in `src/global.d.ts` as it is now (a comment is sufficient). I don't want to have global augmentations in several types, that would be unclear and against the guidelines.

Reply:
Also, I wouldn't mind at all to nuke this typing, as I have **zero** understanding of why these are global methods to begin with. I'd be fine with casting `window` to `any` or _whatever_.

God forbid someone starts using this typing instead of `MemoryOnlyKeys.enable()` / `MemoryOnlyKeys.disable()`, which are also kind-of global, but at least scoped in modules, not in `window`...

Reply:
@cubuspl42 just to confirm we are on the same line, is that how you expect it to look in `global.d.ts`?
```
// Global methods for Onyx key management
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
interface Window {
    enableMemoryOnlyKeys: () => void;
    disableMemoryOnlyKeys: () => void;
}
```

Reply:
@cubuspl42 Also here [the PR](https://github.com/Expensify/App/pull/23831) where this functionality was introduced, maybe it will provide you more context

Reply:
The comment looks fine 👍

I asked [a question on the PR](https://github.com/Expensify/App/pull/23831/files#r1426594284)

Reply:
But if this is actually for manual debugging, I really think that adding any typing might do more harm than good. I can't see any actual usages and can't imagine any scenario when this would be preferred over `MemoryOnlyKeys.enable() / MemoryOnlyKeys.disable()`.

Reply:
@cubuspl42 But without the typing there will be TS errors 
![image](https://github.com/Expensify/App/assets/23176449/2b4f12f9-ee42-43ba-ad90-da541130c22c)

Reply:
Like I mentioned before, I would be fine with a cast too.

```ts
(window as any).whatever = ...
```

This is somewhat hacky typing, but that totally reflects the hacky nature of the code itself.

Reply:
@cubuspl42 Hey, with your solution, we expect `exposeGlobalMemoryOnlyKeysMethods` to look like this? And get rid of Window interface in `global.d.ts`, right?
```
const exposeGlobalMemoryOnlyKeysMethods: ExposeGlobalMemoryOnlyKeysMethods = () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (window as any).enableMemoryOnlyKeys = () => {
        MemoryOnlyKeys.enable();
    };
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (window as any).disableMemoryOnlyKeys = () => {
        MemoryOnlyKeys.disable();
    };
};
```
@fabioh8010 @blazejkustra are you okay with that?

Reply:
@cubuspl42 I'm failing to see how `(window as any).whatever = ...` could be any better than augmenting `Window`. 

First, by doing this we will have to disable `@typescript-eslint/no-explicit-any` rule every time if we use those custom `window` properties because `any` usage is forbidden. Second, this makes the code less safe because in case I need to use the same `window` custom properties in different places I need to be sure I'm writing the correct one, as I don't have type suggestion and type safety anymore by using `(window as any).`.

I know that can be kinda hacky attaching properties to `window`, but this doesn't stop us from adding more type safety to the codebase by augmenting `window` and ensuring this properties are available and correctly typed when used. In a matter of fact, we already had been done for other properties e.g. [here](https://github.com/Expensify/App/blob/main/src/types/modules/pusher.d.ts#L5) and [here](https://github.com/Expensify/App/blob/main/src/types/modules/electron.d.ts#L12).

Regarding the guidelines, we can assure you that we are preparing a PR to make it more explicit that you can not only augment third party modules, but also internal JavaScript ones, if necessary. 🙂

Reply:
> I'm failing to see how (window as any).whatever = ... could be any better than augmenting Window.

As I said, this is not "code" as we typically think about it. It's stuff exposed to the global `window` object so it can be debugged manually via console ([as we now know for sure](https://github.com/Expensify/App/pull/23831/files#r1427398201)).

Someone might call this global thingy instead of proper module-scoped utils if this is typed.

Would you settle on this typing instead?

```ts
// Global methods for Onyx key management for debugging purposes
interface Window {
    set enableMemoryOnlyKeys(value: () => void);
    set disableMemoryOnlyKeys(value: () => void);
}
```

I also added " for debugging purposes" in the suggested comment, as this is now clarified.

Reply:
@cubuspl42 Could you please explain cons of your approach
```
interface Window {
    set enableMemoryOnlyKeys(fn: () => void);
    set disableMemoryOnlyKeys(fn: () => void);
}

```
over the initial one
```
interface Window {
    enableMemoryOnlyKeys: () => void;
    disableMemoryOnlyKeys: () => void;
}
```

Reply:
You meant "pros".

As discussed, these are window-global functions only for manual debugging via the console.

As I repeated multiple times, typing them brings the risk that some engineer uses them in their code, while this is not intended.

With the `set` typing, these utils can only be registered in the `window` object, so they can be called manually from the JS console, but you cannot call them from the TypeScript code without triggering a type error. This reflects the authors' intention. If you were to call this logic from `TypeScript`, you're supposed to use `MemoryOnlyKeys.enable()` / `MemoryOnlyKeys.disable()`;

Reply:
@cubuspl42 Sorry, maybe I did something wrong or just don't understand something, but even with your updates, I'm able to call it from TS files without TS error
![image](https://github.com/Expensify/App/assets/23176449/7631c42d-7e63-41ef-a747-f2d570fe9824)

Reply:
@VickyStash 

You're right, I was confused.

I mean, I don't know why a setter-only thing can be getted and called, but indeed it works as you present.

Let's just add the word "Onyx" to the comment, and we're good, I don't want to stall this PR.

Reply:
@cubuspl42 Updated!

Reply:
Thanks, I'll start testing 👍 If we're far behind main (>200 commits), you can also merge it

Reply:
@cubuspl42 Merged main as well

Reply:
@cubuspl42 FYI latest main has an issue with android, it should be resolved when this [PR](https://github.com/Expensify/App/pull/33224) is merged, I'll monitor it and pull updates to this PR as well

Reply:
Thanks! I tested most flows/platforms, things looks fine. I'll finish when the mentioned issue is resolved 👍